### PR TITLE
Storage refactor

### PR DESCRIPTION
**Pull request by preston-evans98**
This draft breaks the `Storage` trait into two abstraction layers - a
backing store (the db) and an in-memory working set. Working sets are committed
atomically, and provide a read-write abstraction on top of the backing db.

It also makes numerous small changes:

    - Remove ValueReader trait
    - Parameterize all sov-modules-api traits with WorkingSets instead of
      Storage
    - Add "witness" trait, which is produced and consumed by Storage
    - Rename "JmtStorage" to "ProverStorage"
    - Simplify TreeReadLogger
    - Delete ZkTreeDb type
    - Remove usage of FirstReads

### CODE REVIEW

File: demo-app/src/stf.rs [hunk 1]
```diff
@@ -111,6 +122,13 @@ where
         Self::StateRoot,
         Vec<sovereign_sdk::stf::ConsensusSetUpdate<OpaqueAddress>>,
     ) {
-        (jmt::RootHash(self.current_storage.finalize()), vec![])
+        let (cache_log, witness) = self.working_set.borrow_mut().take().unwrap().freeze();
+        let root_hash = self
+            .current_storage
+            .validate_and_commit(cache_log, &witness)
+            .expect("edree update must succed");
+        // let root_hash =
```
Comment (bkolad) [inline demo-app/src/stf.rs:130]: let's remove the comment.

File: sdk/src/state_machine/core/mocks.rs [hunk 2]
```diff
@@ -48,8 +52,35 @@ impl ZkVM for MockZkvm {
...
         _code_commitment: &Self::CodeCommitment,
-    ) -> Result<<<Self as ZkVM>::Proof as crate::zk::traits::ProofTrait<Self>>::Output, Self::Error>
+    ) -> Result<<<Self as ZkVm>::Proof as crate::zk::traits::ProofTrait<Self>>::Output, Self::Error>
     {
         todo!()
     }
 }
+
+#[derive(Default)]
+pub struct MockWitness {
```
Comment (bkolad) [inline sdk/src/state_machine/core/mocks.rs:62]: Should we move out of mocks? Can we make it a default witness?

File: sov-modules/sov-state/src/scratchpad.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+/// A wrapper that adds additional reads and writes on top of an underlying Delta.
+/// These are handly for implementing operations that might revert on top of an existing
+/// working set, without running the risk that the whole working set will be discarded if some particular
+/// operation reverts.
+pub struct RevertableDelta<S: Storage> {
+    inner: Rc<RefCell<Option<Delta<S>>>>,
+    witness: Rc<S::Witness>,
+    cache: Rc<RefCell<StorageInternalCache>>,
+}
+
```
Comment (bkolad) [inline sov-modules/sov-state/src/scratchpad.rs:28]: In the `Witness` we are using a `RefCell `, If we expose it here and move all the `Rc<RefCell<..>>` one level higher and wrap the whole `RevertableDelta => Rc<RefCell< RevertableDelta >>` will it reduce the amount of `borrow & borrow_mut`?

File: sov-modules/sov-state/src/scratchpad.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            .borrow_mut()
+            .merge_reads_left(std::mem::take(&mut self.cache.as_ref().borrow_mut()))
+            .expect("caches must be consistent");
+
+        inner.witness.merge(self.witness.as_ref());
+        inner
+    }
+}
+
+// pub trait WorkingSet: Clone {
```
Comment (bkolad) [inline sov-modules/sov-state/src/scratchpad.rs:175]: Let's remove the commented code.

File: sov-modules/sov-state/src/state_tests.rs [hunk 5]
```diff
@@ -1,122 +1,124 @@
...
+//     let storage = JmtStorage::with_path(&path).unwrap();
+
+//     let mut state_map = StateMap::new(storage.clone(), Prefix::new(vec![0]));
+//     state_map.set(&key, value);
+//     (state_map, storage)
+// }
+
+// #[test]
+// fn test_state_map() {
+//     let path = schemadb::temppath::TempPath::new();
```
Comment (bkolad) [inline sov-modules/sov-state/src/state_tests.rs:86]: Is it possible to activate these tests?

File: demo-app/src/stf.rs (hunk 1 above)
Comment (preston-evans98) [reply demo-app/src/stf.rs:130]: Done

File: sdk/src/state_machine/core/mocks.rs (hunk 2 above)
Comment (preston-evans98) [reply sdk/src/state_machine/core/mocks.rs:62]: Done

**Review by bkolad — APPROVED** [review]

**Pull request merged**